pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -2256,44 +2256,52 @@ public void checkGC() {
-            }
-            CompletableFuture<Void> deleteMetadataFuture = new CompletableFuture<>();
-            getBrokerService().fetchPartitionedTopicMetadataAsync(TopicName.get(topicName.getPartitionedTopicName()))
-                    .thenAccept((metadata -> {
-                        // make sure all sub partitions were deleted
-                        for (int i = 0; i < metadata.partitions; i++) {
-                            if (brokerService.getPulsar().getPulsarResources().getTopicResources()
-                                    .persistentTopicExists(topicName.getPartition(i)).join()) {
-                                throw new UnsupportedOperationException();
-                            }
-                        }
-                    }))
-                    .thenAccept((res) -> partitionedTopicResources.deletePartitionedTopicAsync(topicName)
-                            .thenAccept((r) -> {
-                        deleteMetadataFuture.complete(null);
-                    }).exceptionally(ex -> {
-                        deleteMetadataFuture.completeExceptionally(ex.getCause());
-                        return null;
-                    }))
-                    .exceptionally((e) -> {
-                        if (!(e.getCause() instanceof UnsupportedOperationException)) {
-                            log.error("delete metadata fail", e);
+        PartitionedTopicResources partitionedTopicResources = getBrokerService().pulsar().getPulsarResources()
+                .getNamespaceResources()
+                .getPartitionedTopicResources();
+        if (topicName.isPartitioned()) {
+            return partitionedTopicResources.partitionedTopicExistsAsync(topicName)
+                    .thenCompose(partitionedTopicExist -> {
+                        if (!partitionedTopicExist) {
+                            return CompletableFuture.completedFuture(null);
+                        } else {
+                            return innerDeletePartitionedMetadata(topicName, partitionedTopicResources);
                         }
-                        deleteMetadataFuture.complete(null);
-                        return null;
                     });
-            return deleteMetadataFuture;
-        } catch (Exception e) {
-            return FutureUtil.failedFuture(e);
+        } else {
+            return innerDeletePartitionedMetadata(topicName, partitionedTopicResources);
         }
     }
 
+    private CompletableFuture<Void> innerDeletePartitionedMetadata(TopicName topicName,
+                                                                   PartitionedTopicResources partitionedTopicResources)
+    {
+        return getBrokerService()
+                .fetchPartitionedTopicMetadataAsync(TopicName.get(topicName.getPartitionedTopicName()))
+                .thenCompose((metadata -> {
+                    List<CompletableFuture<Boolean>> persistentTopicExists = new ArrayList<>(metadata.partitions);
```

Comment:
We need a thread safe data structure here.

Reply:
@michaeljmarshall 

This variable doesn't seem to be shared with other threads, ( We are calling ``thenCompose`` not "thenComposeAsync") why does it need to be thread-safe? 
Please give some explanation, Very thanks ~

Reply:
Thanks for your question. My understanding is that `thenCompose` is run on which ever executor completes the `FutureUtil.waitForAll` future, otherwise, `FutureUtil.waitForAll` would be a blocking call. Therefore, line 2280 is potentially on another thread. Let me know what you think.

Reply:
Hi @michaeljmarshall , I think this is ok here.  But please let me know if I mis-understand something.

Reply:
@michaeljmarshall 
Sorry, I missed this comment here.
Very thanks for you.